Add code to handle teardown errors and disruptive tests - IBM Power

gitsridhar opened the pull request:
These changes add to the platform nodes framework to fix teardown
errors and relevant disruptive tests

Signed-off-by: Sridhar Venkat <svenkat@MacBook-Pro-4.attlocal.net>

I have tested the code and removed WIP from the title.

gitsridhar commented:
I am using the following tier4a test case to validate this code:
run-ci -m "tier4a and manage" --ocsci-conf conf/ocsci/production_powervs_upi.yaml --cluster-name ocstest --cluster-path ~ tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py::TestAddCapacityNodeRestart::test_add_capacity_node_restart[1-workload_storageutilization_rbd0]

zmc commented on ocs_ci/ocs/platform_nodes.py:
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
         return default_config_dict
 
 
+class IBMPowerNodes(BaremetalNodes):
```
I'm wondering about the naming and inheritance here. I can't see anything specific to Power, and it seems to me this could be useful for anyone doing an libvirt deployment. Perhaps it ought to be `LibvirtNodes`?

gitsridhar replied:
I put together these changes with the future in mind, for Power currently do things on libvirt, but we certainly would like to do these in PowerVS (IBM Cloud) and Power Baremetal as well one day. When we do that, we may reuse part of Baremetal code that already exists.

gitsridhar replied:
I put together these changes with the future in mind, for Power currently does things on libvirt, but we certainly would like to do these in PowerVS (IBM Cloud) and Power Baremetal as well one day. When we do that, we may reuse part of Baremetal code that already exists.

zmc commented on ocs_ci/ocs/platform_nodes.py:
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
+
+    def ___init__(self):
+        super(IBMPowerNodes, self).__init__()
+        self.powernodes = powernodes.POWERNODES()
```
I don't understand the need for the separate class here; I'd argue that all functionality belongs in one class (even if it should potentially be in a different module and imported into `power_nodes`)

gitsridhar replied:
I see a pattern among other players in this code, they drive changes to OCP from platform_nodes and underlying code is from ocs_ci.utility. I wanted to keep the same structure for this new code as well.

gitsridhar replied:
I see a pattern, changes to OCP infrastructure are driven from platform_nodes, but underlying changes are maintained in utility. I wanted to keep the same structure for this new code.

lukebrowning replied:
> I don't understand the need for the separate class here; I'd argue that all functionality belongs in one class (even if it should potentially be in a different module and imported into `power_nodes`)

Zack, I am not sure I understand your comment.   There are multiple classes today and none of them support libvirt or powervs.   Power actually has two platforms - libvirt and powervs.   There is no standardization around libvirt as each node may have a different VM name.   Power follows this pattern test-<ocp version>-<role>, but another implementation may utilize a different naming scheme.  This is really specific to IBM.  There is no platform independent implementation for libvirt running ocs-ci today, just IBM.   That being said, this code would probably result in an exception on powervs as libvirt is not even configured.  The virsh command is probably not even installed.  So, I think the new class for power needs to import modules for ibm-libvirt and powervs.

zmc replied:
If this is just following existing patterns in the project, I'll just withdraw my comment.

zmc commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,106 @@
+        """
+        Future use
+        """
+        self.return_values = ""
```
If this method isn't doing anything right now, I'd leave it out

zmc reviewed (changes requested):
Nice work! I appreciate the docstrings. I have some comments, and also a question:

How might this affect deployments done with ocs-upi-kvm?

zmc commented:
I see this is still marked as a WIP; is it still being tested or was that just a reflection of my previous review?

lukebrowning commented on ocs_ci/ocs/platform_nodes.py:
```diff
@@ -1685,6 +1686,67 @@ def read_default_config(self, default_config_path):
         return default_config_dict
 
 
+class IBMPowerNodes(NodesBase):
```
Power has two platforms - libvirt and powervs.   This class should be updated to import modules to implement callouts for these two platforms.   libvirt sounds like it mght be standardized, but it is not really.   Each implementation may choose different names.   Power follows this pattern test-ocp-4-6-worker-0, but another implementation may utilize a different naming scheme.   On power, we may implement helper functions that invoke virsh or other functions.   Anyway, none of this would apply to other libvirt platforms as only power implements libvirt ocs-ci.   The powervs module should return not implemented for each callout.  Later, we can plug these with the proper callouts.

gitsridhar replied:
Sure. For now, I am focussing on the libvirt environment. We can take up the PowerVS environment in a separate PR.

lukebrowning replied:
The code still seems like it would try to invoke virsh on powervs.   What am I missing?

gitsridhar commented on ocs_ci/framework/pytest_customization/marks.py:
```diff
@@ -176,7 +176,10 @@
-    config.ENV_DATA["platform"].lower() == "baremetal"
+    (
+        config.ENV_DATA["platform"].lower() == "baremetal"
+        or config.ENV_DATA["platform"].lower() == "powervs"
```
Luke, I will remove this decorator modification. We may not need this and if required I will create a new PR.

lukebrowning replied:
Did you forget to push the change that removes the decorator?

lukebrowning commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,101 @@
+
+        """
+        for node in powernode_machines:
+            result = exec_cmd(f"sudo virsh start test-ocp4-6-{node.name}")
```
You hardcoded the ocp version in the string.   This won't work for other ocp versions.

gitsridhar replied:
I have removed the hardcoded OCP version now in the latest update.

lukebrowning reviewed (changes requested):
Sridhar, you marked my comments as resolved.   Did you forget to push the code changes that resolved them.   You agreed to eliminate the change to the decorator, but the change is still in the latest commit.   The second comment is that your changes donot apply to powervs, but there is nothing to prohibit virsh commands from being invoked on powervs.   These commands are not installed on powervs.   It would be better to have the existing message which is that the function is not implemented, rather than spurious errors like command not found.

gitsridhar commented:
I have added code: 
1. to check for the KVM environment and execute the stop/start/restart functionality. In other environments where virsh is not possible, the code returns an 'unimplemented' exception.
2. Removed hardcoding of OCP version 4-6 and replaced it with a utility API call.

This is in WIP and more testings, I will remove WIP from the title, and the review can be completed.

gitsridhar commented:
Used run-ci -m "tier4a and manage" --ocsci-conf conf/ocsci/production_powervs_upi.yaml --cluster-name ocstest --cluster-path ~ tests/manage/z_cluster/nodes/test_nodes_restart.py::TestNodesRestart::test_nodes_restart[True] 
to test this work.

gitsridhar commented:
I have removed the WIP tag from the title. Please perform a review. @zmc @petr-balogh @vasukulkarni @lukebrowning

gitsridhar commented:
I just published another version - I added IBM Power Platform to ON_PREM in the constants file. I removed 'powervs' from the bare-metal decorator as well.

gitsridhar commented:
@zmc @petr-balogh @vasukulkarni @lukebrowning I made a change based on Luke's comment (move lsmod execution to init method). Please review and merge this code.

I also see Luke's comment asking me to move lsmod to init method is missing now. Strange.

lsmcfadden commented:
@gitsridhar does "I also see Luke's comment asking me to move lsmod to init method is missing now." mean you intend to edit the PR again?

@zmc @petr-balogh @vasukulkarni can you review?  Are you okay to merge it?

lukebrowning reviewed (commented):
Looks good to me.   I have tested this over all of the tier tests.

vasukulkarni commented on ocs_ci/framework/pytest_customization/marks.py:
```diff
@@ -176,7 +176,7 @@
 
 skipif_bm = pytest.mark.skipif(
-    config.ENV_DATA["platform"].lower() == "baremetal"
+    (config.ENV_DATA["platform"].lower() == "baremetal")
```
why was this changed added?

gitsridhar replied:
Vasu, I added code and removed it, so the braces stayed. Let me remove the braces.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+logger = logging.getLogger(__name__)
+
+
+class POWERNODES(object):
```
Class names should be PascalCase, not all caps.

```suggestion
class PowerNodes(object):
```

gitsridhar replied:
Completed in the latest push.

clacroix12 commented on ocs_ci/ocs/platform_nodes.py:
```diff
@@ -1712,6 +1713,102 @@ def read_default_config(self, default_config_path):
+                "This is not libvirt environment. Start nodes not implemented"
+            )
+
+    def restart_nodes(self, nodes, timeout=540, wait=True, force=True):
```
`timeout` and `wait` args are not used or included in the docstring.

gitsridhar replied:
Completed with latest push.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+    Wrapper for PowerNodes
+    """
+
+    isKVM = False
```
Does this need to be defined here or can we stick to initializing in the `__init__`?

gitsridhar replied:
Completed with the latest push. Removed global entry.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+
+    def iskvm(self):
+        """
+        Verify if this environment is libvirt or kvm.
+        Args: None
+        Returns:
+            bool: True if this is kvm environment, False otherwise
```
```suggestion
        Verify if this environment is libvirt or kvm.
        
        Returns:
            bool: True if this is kvm environment, False otherwise

```

gitsridhar replied:
Done.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+
+    def restart_powernodes_machines(self, powernode_machines, force=True):
+        """
+
```
No need for the empty line at the start of the docstring

```suggestion
```

gitsridhar replied:
Done.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        ocpversion = get_ocp_version("-")
+        for node in powernode_machines:
+            cmd = "sudo virsh shutdown test-ocp" + ocpversion + f"-{node.name}"
```
We shouldn't be using concatenation and f-strings together like this. Let's simplify and make the whole thing an f-string.
 
```suggestion
            cmd = f"sudo virsh shutdown test-ocp{ocpversion}-{node.name}"
```

gitsridhar replied:
Done.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+        for node in powernode_machines:
+            cmd = "sudo virsh shutdown test-ocp" + ocpversion + f"-{node.name}"
+            result = exec_cmd(cmd)
+            """
+            if result.returncode == 0:
+                return False
+            else
+                return True
+            """
```
Does this belong here?

gitsridhar replied:
Removed commented code.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,132 @@
+        ocpversion = get_ocp_version("-")
+        for node in powernode_machines:
+            result = exec_cmd(
+                "sudo virsh start test-ocp" + ocpversion + f"-{node.name}"
```
Same thing here, just use one f-string:

```suggestion
                f"sudo virsh start test-ocp{ocpversion}-{node.name}"
```

gitsridhar replied:
Done.

gitsridhar commented:
@clacroix12 I took care of your review comments, after more tests, I will remote WIP tag.

gitsridhar commented:
Test complete, I removed WIP from the title. Please review.

clacroix12 commented on ocs_ci/utility/powernodes.py:
```diff
@@ -0,0 +1,135 @@
+logger = logging.getLogger(__name__)
+
+
+class POWERNodes(object):
```
Must have missed this one in my initial review. This class should be `PowerNodes`.

gitsridhar replied:
Sure. I will submit a new PR later to rename this class.

clacroix12 reviewed (approved):
Just the one issue that can be taken up in a follow-up PR.

lukebrowning commented:
@vasukulkarni Please merge this.
